packages/notebook/src/searchprovider.ts

```diff
@@ -562,44 +662,142 @@ export class NotebookSearchProvider extends SearchProvider<NotebookPanel> {
-          provider.setIsActive(
-            this.widget.content.isSelectedOrActive(cells[index])
-          )
-        )
+  private _stopObservingLastCell() {
+    if (this._editorSelectionsObservable) {
+      this._editorSelectionsObservable.changed.disconnect(
+        this._setSelectedLines,
+        this
       );
+    }
+  }
+
+  private _setSelectedLines() {
+    const editor = this.widget.content.activeCell?.editor;
+    if (!editor) {
+      return;
+    }
+
+    const selection = editor.getSelection();
+    const { start, end } = selection;
+
+    const newLines =
+      end.line === start.line && end.column === start.column
+        ? 0
+        : end.line - start.line + 1;
+
+    this._textSelection = selection;
+
+    if (newLines !== this._selectedLines) {
+      this._selectedLines = newLines;
+      this._updateSelectionMode();
+    }
+    this._filtersChanged.emit();
+  }
+
+  private _textSelection: CodeEditor.IRange | null = null;
+
+  /**
+   * Set whether the engines should search within selection only or full text.
+   */
+  private async _setEnginesSelectionSearchMode() {
+    let textMode: boolean;
+
+    if (!this._onSelection) {
+      // When search in selection is off we always search full text
+      textMode = false;
+    } else {
+      // When search in selection is off we either search in full cells
+      // (toggling off isActive flag on search enginges of non-selected cells)
+      // or in selected text of the active cell.
+      textMode = this._selectionSearchMode === 'text';
+    }
 
-      this._onSearchProviderChanged();
+    if (this._selectionLock) {
+      return;
     }
+
+    // Clear old selection restrictions or if relevant, set current restrictions for active provider.
+    await Promise.all(
+      this._searchProviders.map((provider, index) => {
+        const isCurrent = this.widget.content.activeCellIndex === index;
+        return provider.setSearchSelection(
+          isCurrent && textMode ? this._textSelection : null
+        );
+      })
+    );
+  }
+
+  private async _onCellSelectionChanged() {
+    const cells = this.widget.content.widgets;
+    let selectedCells = 0;
+    await Promise.all(
+      cells.map(async (cell, index) => {
+        const provider = this._searchProviders[index];
+        const isSelected = this.widget.content.isSelectedOrActive(cell);
+        if (isSelected) {
+          selectedCells += 1;
+        }
+        if (provider && this._onSelection) {
+          await provider.setIsActive(isSelected);
+        }
+      })
+    );
+
+    if (selectedCells !== this._selectedCells) {
+      this._selectedCells = selectedCells;
+      this._updateSelectionMode();
+    }
+
+    this._filtersChanged.emit();
   }
 
   private _currentProviderIndex: number | null = null;
   private _filters: IFilters | undefined;
-  private _onSelectedCells = false;
+  private _onSelection = false;
+  private _selectedCells: number = 1;
+  private _selectedLines: number = 0;
```

Comment:
This is not always correctly updated (please see GIF with testing in the main comment)